File: exercise2.cpp
#include <iostream>
#include <ctime>
#include <cstdlib>

#define SIZE 5
#define MINES 5

int board[SIZE][SIZE];
bool revealed[SIZE][SIZE];

void init() {
    srand(time(0));
    for(int i = 0; i < MINES; i++) {
        int x = rand() % SIZE;
        int y = rand() % SIZE;
        if(board[x][y] != -1) {
            board[x][y] = -1;
        } else {
            i--;
        }
    }
}

int countMines(int x, int y) {
    int count = 0;
    for(int i = -1; i <= 1; i++) {
        for(int j = -1; j <= 1; j++) {
            if(x+i >= 0 && x+i < SIZE && y+j >= 0 && y+j < SIZE && board[x+i][y+j] == -1) {
                count++;
            }
        }
    }
    return count;
}

void reveal(int x, int y) {
    if(x >= 0 && x < SIZE && y >= 0 && y < SIZE && !revealed[x][y]) {
        revealed[x][y] = true;
        if(board[x][y] == -1) {
            std::cout << "Game Over\n";
            exit(0);
        } else {
            board[x][y] = countMines(x, y);
        }
    }
}

void printBoard() {
    for(int i = 0; i < SIZE; i++) {
        for(int j = 0; j < SIZE; j++) {
            if(revealed[i][j]) {
                if(board[i][j] == -1) {
                    std::cout << "* ";
                } else {
                    std::cout << board[i][j] << " ";
                }
            } else {
                std::cout << "- ";
            }
        }
        std::cout << "\n";
    }
}

int main() {
    init();
    while(true) {
        printBoard();
        int x, y;
        std::cout << "Enter row and column: ";
        std::cin >> x >> y;
        reveal(x, y);
    }
    return 0;
}

File: exercises/python/exercise2.py
import random

SIZE = 5
MINES = 5

board = [[0 for _ in range(SIZE)] for _ in range(SIZE)]
revealed = [[False for _ in range(SIZE)] for _ in range(SIZE)]


def init():
    mines_placed = 0
    while mines_placed < MINES:
        x = random.randint(0, SIZE - 1)
        y = random.randint(0, SIZE - 1)
        if board[x][y] != -1:
            board[x][y] = -1
            mines_placed += 1


def count_mines(x, y):
    count = 0
    for i in range(-1, 2):
        for j in range(-1, 2):
            if 0 <= x + i < SIZE and 0 <= y + j < SIZE and board[x + i][y + j] == -1:
                count += 1
    return count


def reveal(x, y):
    if 0 <= x < SIZE and 0 <= y < SIZE and not revealed[x][y]:
        revealed[x][y] = True
        if board[x][y] == -1:
            print("Game Over")
            exit(0)
        else:
            board[x][y] = count_mines(x, y)


def print_board():
    for i in range(SIZE):
        for j in range(SIZE):
            if revealed[i][j]:
                if board[i][j] == -1:
                    print("*", end=" ")
                else:
                    print(board[i][j], end=" ")
            else:
                print("-", end=" ")
        print()


def main():
    init()
    while True:
        print_board()
        x, y = map(int, input("Enter row and column: ").split())
        reveal(x, y)


if __name__ == "__main__":
    main()
